tests/storage/test_raft_storage.rs

```diff
@@ -146,43 +163,77 @@ fn test_engine_leader_change_twice() {
-    let mut engine = util::BlockEngine::new(engine);
+    let region0 = cluster.get_region(b"");
+    let peers = region0.get_peers();
+    cluster.must_transfer_leader(region0.get_id(), peers[0].clone());
     let config = Config::default();
-    let mut storage = Storage::from_engine(engine.clone(), &config).unwrap();
-    storage.start(&config).unwrap();
-    let mut ctx = Context::new();
-    ctx.set_region_id(region.get_id());
-    ctx.set_region_epoch(region.get_region_epoch().clone());
-    ctx.set_peer(peers[0].clone());
-    let (tx, rx) = channel();
+
+    let engine0 = cluster.sim.rl().storages[&peers[0].get_id()].clone();
+    let mut engine0 = util::BlockEngine::new(engine0);
+    let mut storage0 = Storage::from_engine(engine0.clone(), &config).unwrap();
+    storage0.start(&config).unwrap();
+
+    let mut ctx0 = Context::new();
+    ctx0.set_region_id(region0.get_id());
+    ctx0.set_region_epoch(region0.get_region_epoch().clone());
+    ctx0.set_peer(peers[0].clone());
+    let (prewrite_tx, prewrite_rx) = channel();
     let (stx, srx) = channel();
-    engine.block_snapshot(stx.clone());
-    storage.async_prewrite(ctx.clone(),
+    engine0.block_snapshot(stx.clone());
+    storage0.async_prewrite(ctx0,
                         vec![Mutation::Put((make_key(b"k"), b"v".to_vec()))],
                         b"k".to_vec(),
                         10,
                         Options::default(),
                         box move |res: storage::Result<_>| {
-            if let storage::Error::Engine(engine::Error::Request(ref e)) = *res.as_ref()
-                .err()
-                .unwrap() {
-                assert!(e.has_stale_command());
-            } else {
-                panic!("expect stale command, but got {:?}", res);
+            match res {
+                Err(storage::Error::Engine(engine::Error::Request(ref e))) => {
+                    assert!(e.has_stale_command());
+                    prewrite_tx.send(false).unwrap();
+                }
+                Ok(_) => {
+                    prewrite_tx.send(true).unwrap();
+                }
+                _ => {
+                    panic!("expect stale command, but got {:?}", res);
+                }
             }
-            tx.send(1).unwrap();
         })
         .unwrap();
     // wait for the message, the prewrite should be blocked at snapshot stage.
     srx.recv_timeout(Duration::from_secs(2)).unwrap();
     // Transfer leader twice, then unblock snapshot.
-    cluster.must_transfer_leader(region.get_id(), peers[1].clone());
-    cluster.must_transfer_leader(region.get_id(), peers[0].clone());
-    engine.unblock_snapshot();
-
-    rx.recv_timeout(Duration::from_secs(5)).unwrap();
+    cluster.must_transfer_leader(region0.get_id(), peers[1].clone());
+    cluster.must_transfer_leader(region0.get_id(), peers[0].clone());
+    engine0.unblock_snapshot();
+
+    // the snapshot request may meet read index, scheduler will retry the request.
```

Comment:
Yes. If the it meets read index and prewrite successes via retry, the key can also commit successfully on the other peer.  https://github.com/pingcap/tikv/pull/2111/files#diff-e61719323a839fa832f131322cec829aR225